agent/src/main/java/com/code_intelligence/jazzer/autofuzz/AutofuzzCodegenVisitor.java

```diff
@@ -0,0 +1,112 @@
+// Copyright 2021 Code Intelligence GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.code_intelligence.jazzer.autofuzz;
+
+import java.util.Stack;
+import java.util.stream.Collectors;
+
+public class AutofuzzCodegenVisitor {
+  private final Stack<Group> groups = new Stack<>();
+  private int variableCounter = 0;
+
+  AutofuzzCodegenVisitor() {
+    init();
+  }
+
+  private void init() {
+    pushGroup("", "", "");
+  }
+
+  public void pushGroup(String prefix, String delimiter, String suffix) {
+    groups.push(new Group(prefix, delimiter, suffix));
+  }
+
+  public void pushElement(String element) {
+    groups.peek().push(element);
+  }
+
+  public void popElement() {
+    groups.peek().pop();
+  }
+
+  public void popGroup() {
+    if (groups.size() == 1) {
+      throw new AutofuzzError("popGroup must be called exactly once for every pushGroup");
+    }
+    pushElement(groups.pop().toString());
+  }
+
+  public String generate() {
+    if (groups.size() != 1) {
+      throw new AutofuzzError("popGroup must be called exactly once for every pushGroup");
```

Comment:
@fmeum I've just started with the review and give the new functionality a try with commons imaging (` --autofuzz="org.apache.commons.imaging.Imaging::getBufferedImage"`) and this triggers this exception in some runs.

Reply:
I pushed two new commits that should fix this. I let it run for a while it did not encounter any more `AutofuzzError`s.